src/vs/platform/quickinput/browser/quickPickPin.ts

```diff
@@ -0,0 +1,86 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { Codicon } from 'vs/base/common/codicons';
+import { localize } from 'vs/nls';
+import { ServicesAccessor } from 'vs/platform/instantiation/common/instantiation';
+import { IQuickPick, IQuickPickItem, IQuickPickItemButtonEvent, QuickPickItem } from 'vs/platform/quickinput/common/quickInput';
+import { IStorageService, StorageScope, StorageTarget } from 'vs/platform/storage/common/storage';
+import { ThemeIcon } from 'vs/platform/theme/common/themeService';
+
+/**
+ * Initially, adds pin buttons to all @param quickPick items.
+ * When pinned, a copy of the item will be moved to the end of the pinned list and any duplicate within the pinned list will
+ * be removed. Pin and pinned button events trigger updates to the underlying storage.
+ * The @param callback to run after the items have been formatted or
+ * when one the pinned list is updated
+ */
+export function formatPinnedItems(accessor: ServicesAccessor, storageKey: string, quickPick: IQuickPick<IQuickPickItem>, callback: () => Promise<void>): void {
+	const storageService = accessor.get(IStorageService);
+	quickPick.onDidTriggerItemButton((e) => {
+		if (e.button.iconClass === ThemeIcon.asClassName(Codicon.pin) || e.button.iconClass === ThemeIcon.asClassName(Codicon.pinned)) {
+			_formatPinnedItems(storageKey, quickPick, storageService, e, callback);
+			callback();
```

Comment:
So if I'm reading this right... the `callback()` gets called if you pin or unpin an item. Based on your call of `formatPinnedItems` you call `this.runRecent(type, filterMode, value)`... Is that the expected behavior? I would think that pinning/unpinning would _only_ cause the pinning stuff to happen, not anything else.

Reply:
yeah you're reading it correctly - when I instead call `quickPick.show()` there it doesn't work right.

Reply:
Updating the `.items` of the quick pick object should refresh the UI

Reply:
it does but adds the pinned category again to the list, so it's duplicated

Reply:
and in formatPinnedItems, I recreate the list of items every time - don't just push or remove ones from it

Reply:
looking into it

Reply:
You _do_ need to recreate the list. The quickpick will only update when `.items` is set to something else. The list can't be mutated in place.

Reply:
agree it's odd - don't really understand how it's possible